### ISSUE TITLE
PySpark DataFrame.rdd incorrectly inferred as Column

### ISSUE THREAD

**Issue by andyli-oai** [astral-sh/ty#4286]
### Summary

With full PySpark installed, ty incorrectly reports `call-non-callable` when calling `DataFrame.rdd.toLocalIterator()`.

```python
from typing import reveal_type

from pyspark.sql import DataFrame


def read_rows(dataframe: DataFrame) -> None:
    reveal_type(dataframe.rdd)  # ty: RDD[Row] | Column; mypy: RDD[Row]
    dataframe.rdd.toLocalIterator(prefetchPartitions=True)  # ty: call-non-callable
```

### Expected behavior

ty should infer `DataFrame.rdd` as `RDD[Row]` and accept `toLocalIterator()` when full PySpark is installed and `is_remote_only()` returns `False`.

### Cause

PySpark conditionally defines `rdd` and implements a dynamic attribute fallback:

```python
class DataFrame:
    if not is_remote_only():

        @property
        def rdd(self) -> RDD[Row]:
            ...

    def __getattr__(self, name: str) -> Column:
        ...
```

ty combines the property and fallback into `RDD[Row] | Column`. Consequently, it treats `toLocalIterator` as either a bound method or a non-callable `Column`, even though the `rdd` property exists in the installed environment.

Possibly related: #1480.

### Version
```
ty 0.0.72
Python 3.12.13
PySpark 4.0.2 and 4.2.0
```

**Comment by carljm** [comment]
Here's a simpler self-contained version of this issue:

```py
def condition() -> bool:
    return False

class Example:
    if condition():
        value: int = 1

    def __getattr__(self, name: str) -> str:
        return name

reveal_type(Example().value)
```

Other type checkers ignore the fact that `value` is only conditionally defined, and treat it as always existing, inferring a type of `int`. ty considers the case that `condition()` might be false, and thus that `__getattr__` would control the type of `value`.

To me, ty's behavior here seems more correct; I'm not sure what the general rationale would be for assuming that all conditionals in a class body are always true.

Actually inferring `is_remote_only()` to be true based on installed context doesn't seem very feasible; it would require deep inspection of the internals of that function.

This seems like more of a problem with PySpark's full-vs-remote-only model and type stubs, that would ideally be fixed on the PySpark side.

I guess we could decide for compatibility to follow other type checkers and assume conditionally-defined attributes fully define the type of the attribute, without fallback to `__getattr__`.

**Comment by tigerhawkvok** [comment]
> Other type checkers ignore the fact that value is only conditionally defined, and treat it as always existing, inferring a type of int. ty considers the case that condition() might be false, and thus that __getattr__ would control the type of value.
> 
> To me, ty's behavior here seems more correct [...]

I strongly disagree. ty's behaviour here is more wrong -- `bool` is a superset of return values actually provided, but it is impossible for `condition()` to return anything other than `False`. In deferring to the nebulous "maybe the declared return type is because the interior code may change one day" ty breaks facts-on-the-ground based on what other libraries the dev has no control over provides. 

Perhaps a `strict-return-type` setting could toggle the two behaviours; but OP has no control over what Apache drops in Pyspark. Boilerplate assertions and type castings just muddy up code behaviour when, by restricting the inferred type of `condition()`, if `condition()` ever changed, errors would immediately show up -- no harm, no foul, and meanwhile keeping _improved_ utility.

**Comment by carljm** [comment]
@tigerhawkvok I'm confused by your reply.

In my simplified example above, `condition()` in fact always returns `False`. This means that at runtime, `value: int = 1` will never execute, `__getattr__` will always be called, and the correct inference is `str`. All other type checkers are blindly assuming here that `condition()` is always truthy, and they all give a definitely-exactly-wrong result in this example by inferring `int`. Only ty gives an inference (`int | str`) that even acknowledges the possibility of the correct result of `str`. This is why I suggest that ty's inference is more correct. That seems unequivocally true, in this example at least.

No type checker looks inside the `condition` function to see the actual return type and refine the annotated return type, which is the expectation it sounds like you might have? You could argue that they should, but this quickly becomes infeasible in non-trivial real-world cases, including [the actual implementation](https://github.com/apache/spark/blob/v4.1.2/python/pyspark/util.py#L852-L890) of `is_remote_only`, which depends, among other things, on external environment variable state and a modifiable global variable.

So the question here is simply, should type checkers assume that conditions inside a class body are always true (that is, all declarations and assignments are always reachable) and therefore confidently give more precise answers that are sometimes precisely wrong? Or should they (as ty does) give an answer that considers both possibilities?
